pkg/service/volume_service_test.go

```diff
@@ -355,6 +355,97 @@ func (suite *VolumeServiceTestSuite) TestGetVolumes() {
 	}
 
 }
+func (suite *VolumeServiceTestSuite) TestVolumeSortFilters() {
+	// sort unit test
+	sfilter := new(param.GetParams)
+	var sortOrderList []param.SortOrder
+	sortOrderList = append(sortOrderList, param.SortOrder{
+		Field:     *nimbleos.VolumeFields.Name,
+		Ascending: true,
+	})
+
+	sfilter.WithSortBy(sortOrderList)
+	volumes, err := suite.volumeService.GetVolumes(sfilter)
+	if err != nil || len(volumes) == 0 {
+		suite.T().Errorf("TestGetVolumes(): Unable to fetch volumes, err: %v", err.Error())
+		return
+	}
+
+}
+
+func (suite *VolumeServiceTestSuite) TestVolumeSearchFilters() {
+	// search filter unit test
+	sfilter := &param.GetParams{}
+	sf := &param.SearchFilter{
+		FieldName: nimbleos.VolumeFields.Name,
+		Operator:  param.EQUALS.String(),
+		Value:     "GetVolume",
+	}
+
+	sfilter.WithSearchFilter(sf)
+	volumes, _ := suite.volumeService.GetVolumes(sfilter)
+	if len(volumes) == 0 {
+		suite.T().Errorf("TestGetVolumes(): Unable to fetch volumes")
```

Comment:
Can we test that the request was sent with the correct search filters?  Checking that some volumes are returned only tells us that a valid request was sent, not that it was the correct request.

These test cases are billed as unit tests, but actually they act as integration tests.  A unit test does not depend upon the behavior of another device on the network.  Could we mock the resty client to control the behavior for unit testing?

It might not be correct to fail the test when len(volumes) == 0, depending on the state of the array.  Does it only succeed if the array has a volume named "GetVolume"?

Reply:
Agree. we shouldnt check on len(volumes) == 0. Actually, I should check on error instead of len(volumes) since that will validate the filter . An incorrect search filter will result in an error. 
This is unit test to validate the correctly implementation of search , sort etc .

Reply:
I disagree.  The SDK could, for example, omit the filter entirely from the request and this test case would still pass.

You are calling it a "unit test", but do you think that is what it is?

Reply:
Disagree! 
the search filter is optional to request , here it is actually validating that it works. Since generated endpoints are not the same with and without filters to request. 
It also serves as an example for users, how to use search filters,s, or sorting with SDK.

Reply:
How does this test validate the generated endpoint?

Reply:
With endpoints, filter arguments are passed as key/value which gets validated on Array . if there is an error test will fail. hope this helps.

Reply:
I think you are evaluating the test case based on your knowledge of the implementation, and I expect you are correct given the current implementation.  But I bet I could insert a bug into the implementation that causes the filters not to work at all, and this test case wouldn't catch it.

Reply:
Sorry I'm late to the conversation.  I believe @rkumpf is pointing out that we don't validate any of the possible inputs to the search filter.  For example, the operator could be something like "square_root_of" which the server doesn't support.  And the name could be something that fails to be encode correctly.  I think it's a good practice to validate the inputs to the request, and not just validate that you get something back when the request is correctly formed.

Reply:
Yes, there is no validation on search filter values we could enhance it to have validation in place. But still, the even if user supply incorrect values to search filter, the Array would throw an exception, which is returned to the caller in err form.

Reply:
Sure. But again, that tests the array REST API server; not your code.

Reply:
isn't indirectly my code getting validated :)